### PR TITLE
Structure and first pass for grouping validation errors

### PR DESCRIPTION
**Pull request by anissa-agahchen**
# Overview

## Links to jira tickets

- {Include links to all of the applicable Jira tickets}

## This PR contains the following changes

- {List all the relevant changes. If there are many changes across many Jira tickets, organize the changes by Jira ticket}

## This PR contains the following types of changes

- [ ] New feature (change which adds functionality)
- [ ] Enhancement (improvements to existing functionality)
- [ ] Bug fix (change which fixes an issue)
- [ ] Misc cleanup / Refactoring / Documentation
- [ ] Version change

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [ ] I have performed a self-review of my own code

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

Please add any relevant UI screenshots, if applicable.

### CODE REVIEW

File: api/src/paths/dwc/validate.ts [hunk 1]
```diff
@@ -211,7 +211,7 @@ export function persistParseErrors(): RequestHandler {
         connection
       );
 
-      await insertSubmissionMessage(submissionStatusId, 'Error', parseError, connection);
+      await insertSubmissionMessage(submissionStatusId, 'Error', parseError, 'miscelaneous', connection);
```
Comment (sdevalapurkar) [inline api/src/paths/dwc/validate.ts]: spelling here:
```suggestion
      await insertSubmissionMessage(submissionStatusId, 'Error', parseError, 'miscellaneous', connection);
```

File: api/src/paths/dwc/validate.ts [hunk 2]
```diff
@@ -418,9 +423,15 @@ export const insertSubmissionMessage = async (
   submissionStatusId: number,
   submissionMessageType: string,
   message: string,
+  error_code: string,
```
Comment (sdevalapurkar) [inline api/src/paths/dwc/validate.ts]: nit, but for consistency can we make this camelCase?
```
  errorCode: string,
```

File: api/src/paths/project/{projectId}/survey/{surveyId}/observation/submission/get.ts [hunk 3]
```diff
@@ -162,7 +159,7 @@ export function getOccurenceSubmission(): RequestHandler {
 
       return res.status(200).json(getOccurrenceSubmissionData);
     } catch (error) {
-      defaultLog.debug({ label: 'getOccurenceSubmission', message: 'error', error });
+      defaultLog.debug({ label: 'getOccurrenceSubmission', message: 'error', error });
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observation/submission/get.ts:162]: the actual function name on line 94 needs to be changed to fix the spelling as well

File: api/src/utils/media/csv/csv-file.ts [hunk 4]
```diff
@@ -287,21 +287,29 @@ export class XLSXCSV implements IWorkbook {
...
+  | 'duplicate_header'
+  | 'unknown_header'
+  | 'missing_required_header'
+  | 'missing_recommended_header'
+  | 'miscellaneous_for_header';
+
+export type IRowErrorCode =
+  | 'missing_required_field'
+  | 'unexpected_formats'
+  | 'Values_exceeded'
```
Comment (sdevalapurkar) [inline api/src/utils/media/csv/csv-file.ts]: nit for consistency:
```suggestion
  | 'values_exceeded'
```

File: api/src/utils/media/csv/validation/csv-header-validator.test.ts [hunk 5]
```diff
@@ -87,19 +85,19 @@ describe('hasRequiredHeadersValidator', () => {
 
     expect(csvWorkSheet.csvValidation.headerErrors).to.eql([
       {
-        code: 'MissingRequiredHeader',
+        coerror_codede: 'MissingRequiredHeader',
```
Comment (sdevalapurkar) [inline api/src/utils/media/csv/validation/csv-header-validator.test.ts]: should this say:
```suggestion
        error_code: 'MissingRequiredHeader',
```

File: app/src/features/surveys/view/SurveyObservations.tsx [hunk 6]
```diff
@@ -186,6 +193,42 @@ const SurveyObservations = () => {
     </IconButton>
   );
 
+  const groupType = {
+    mandatory: 'Mandatory fields have not been filled out in your file',
+    unsupported_header: 'Column headers in your file are not supported',
+    miscellaneous: 'Miscellaneous'
```
Comment (sdevalapurkar) [inline app/src/features/surveys/view/SurveyObservations.tsx]: should this say something along the lines of:
```suggestion
    miscellaneous: 'Miscellaneous errors in your file'
```

File: app/src/features/surveys/view/SurveyObservations.tsx [hunk 7]
```diff
@@ -186,6 +193,42 @@ const SurveyObservations = () => {
...
+  const messageList = submissionStatus?.messages;
+
+  const messageObject =
+    (messageList &&
+      messageList.reduce((workingData: any, row: any) => {
+        let group;
+
+        switch (row.error_code) {
+          case 'missing_required_header':
+          case 'missing_required_field':
```
Comment (sdevalapurkar) [inline app/src/features/surveys/view/SurveyObservations.tsx]: question:
is this how you do it if multiple cases have the same logic? or can we remove line 210?

File: app/src/features/surveys/view/SurveyObservations.tsx [hunk 8]
```diff
@@ -186,6 +193,42 @@ const SurveyObservations = () => {
...
+            group = 'unsupported_header';
+            break;
+          default:
+            group = 'miscellaneous';
+        }
+
+        let groupArray = workingData[group];
+
+        if (!groupArray) {
+          workingData[group] = [row.message];
```
Comment (sdevalapurkar) [inline app/src/features/surveys/view/SurveyObservations.tsx]: can this just be:
```suggestion
          groupArray = [row.message];
```

File: app/src/features/surveys/view/SurveyObservations.tsx [hunk 9]
```diff
@@ -186,6 +193,42 @@ const SurveyObservations = () => {
...
+            group = 'mandatory';
+            break;
+          case 'unknown_header':
+            group = 'unsupported_header';
+            break;
+          default:
+            group = 'miscellaneous';
+        }
+
+        let groupArray = workingData[group];
```
Comment (sdevalapurkar) [inline app/src/features/surveys/view/SurveyObservations.tsx]: this entire logic maay be less confusing without this assignment to a variable happening

File: app/src/features/surveys/view/SurveyObservations.tsx (hunk 7 above)
Comment (anissa-agahchen) [reply app/src/features/surveys/view/SurveyObservations.tsx]: yeah ... so this is to group the errors generated under required fields/headers under the mandatory grouping

File: app/src/features/surveys/view/SurveyObservations.tsx (hunk 6 above)
Comment (anissa-agahchen) [reply app/src/features/surveys/view/SurveyObservations.tsx]: ok.  Just FYI though ... the miscellaneous is to bucket things that are not assigned a grouping yet.  Not sure if we'll keep this concept longterm .. but for now it's helpful

File: app/src/features/surveys/view/SurveyObservations.tsx (hunk 8 above)
Comment (anissa-agahchen) [reply app/src/features/surveys/view/SurveyObservations.tsx]: @sdevalapurkar can you take a look again?

File: app/src/features/surveys/view/SurveyObservations.tsx (hunk 6 above)
Comment (sdevalapurkar) [reply app/src/features/surveys/view/SurveyObservations.tsx]: cool cool

File: app/src/features/surveys/view/SurveyObservations.tsx (hunk 7 above)
Comment (sdevalapurkar) [reply app/src/features/surveys/view/SurveyObservations.tsx]: right - but can you group multiple case statements like this?

File: app/src/features/surveys/view/SurveyObservations.tsx (hunk 8 above)
Comment (sdevalapurkar) [reply app/src/features/surveys/view/SurveyObservations.tsx]: yea I think this is easier to understand

File: app/src/features/surveys/view/SurveyObservations.tsx (hunk 7 above)
Comment (anissa-agahchen) [reply app/src/features/surveys/view/SurveyObservations.tsx]: yeah, it's a nice feature :-)

File: api/src/paths/project/{projectId}/survey/{surveyId}/observation/submission/get.test.ts [hunk 10]
```diff
@@ -118,10 +118,14 @@ describe('getObservationSubmission', () => {
 
     sinon.stub(survey_occurrence_queries, 'getLatestSurveyOccurrenceSubmissionSQL').returns(SQL`something`);
 
-    const result = observationSubmission.getOccurenceSubmission();
+    const result = observationSubmission.getOccurrenceSubmission();
+
+    console.log('result is: ', result);
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observation/submission/get.test.ts]: remove this
```suggestion
```

File: api/src/paths/project/{projectId}/survey/{surveyId}/observation/submission/get.test.ts [hunk 11]
```diff
@@ -118,10 +118,14 @@ describe('getObservationSubmission', () => {
...
 
     await result(sampleReq, sampleRes as any, (null as unknown) as any);
 
+    console.log('actual result is: ', actualResult);
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observation/submission/get.test.ts]: and this:
```suggestion
```

File: app/src/features/surveys/view/SurveyObservations.tsx [hunk 12]
```diff
@@ -186,6 +193,40 @@ const SurveyObservations = () => {
     </IconButton>
   );
 
+  const groupType = {
+    mandatory: 'Mandatory fields have not been filled out in your file',
+    unsupported_header: 'Column headers in your file are not supported',
+    miscellaneous: 'Miscellaneous errors in your file'
```
Comment (sdevalapurkar) [inline app/src/features/surveys/view/SurveyObservations.tsx]: supernit:
```suggestion
    miscellaneous: 'Miscellaneous errors exist in your file'
```

**Comment by sdevalapurkar** [comment]
overall code looks good

**Review by NickPhura — APPROVED** [review]

**Pull request merged**